redshift/redshift.go

```diff
@@ -49,68 +87,76 @@ func NewRedshift(host, port, db, user, pwd string, timeout int, s3Info S3Info) (
-func (r *Redshift) CopyJSONDataFromS3(schema, table, file, jsonpathsFile string) error {
-	creds := fmt.Sprintf(`CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'`, r.s3Info.AccessID, r.s3Info.SecretKey)
+// GetJSONCopySQL copies JSON data present in an S3 file into a redshift table.
+// if not using jsonPaths, set to "auto"
+func (r *Redshift) GetJSONCopySQL(schema, table, filename, jsonPaths string, creds, gzip bool) string {
+	credSQL := ""
+	if creds {
+		credSQL = fmt.Sprintf(`CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'`, r.s3Info.AccessID, r.s3Info.SecretKey)
+	}
+	gzipSQL := ""
+	if gzip {
+		gzipSQL = "GZIP"
+	}
 	copyCmd := fmt.Sprintf(
-		`COPY "%s"."%s" FROM '%s' WITH json '%s' region '%s' timeformat 'epochsecs' COMPUPDATE ON %s`,
-		schema, table, file, jsonpathsFile, r.s3Info.Region, creds,
-	)
-	_, err := r.logAndExec(copyCmd)
-	return err
+		`COPY "%s"."%s" FROM '%s' WITH %s JSON '%s' REGION '%s' TIMEFORMAT 'auto' COMPUPDATE ON %s`,
+		schema, table, filename, gzipSQL, jsonPaths, r.s3Info.Region, credSQL)
+	return copyCmd
 }
 
-// CopyGzipCsvDataFromS3 copies gzipped CSV data from an S3 file into a redshift table.
-func (r *Redshift) CopyGzipCsvDataFromS3(schema, table, file string, ts postgres.TableSchema, delimiter rune) error {
-	cols := []string{}
-	sort.Sort(ts)
-	for _, ci := range ts {
-		cols = append(cols, ci.Name)
+// GetCSVCopySQL copies gzipped CSV data from an S3 file into a redshift table.
+func (r *Redshift) GetCSVCopySQL(schema, table, file string, ts Table, delimiter rune, creds, gzip bool) string {
+	credSQL := ""
+	if creds {
+		credSQL = fmt.Sprintf(`CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'`, r.s3Info.AccessID, r.s3Info.SecretKey)
+	}
+	gzipSQL := ""
+	if gzip {
+		gzipSQL = "GZIP"
+	}
+
+	cols := SortableColumns{}
+	cols = append(cols, ts.Columns...)
+	sort.Sort(cols)
+	colStrings := []string{}
+	for _, ci := range cols {
+		colStrings = append(colStrings, ci.Name)
 	}
 	copyCmd := fmt.Sprintf(
-		`COPY "%s"."%s" (%s) FROM '%s' WITH REGION '%s' GZIP CSV DELIMITER '%c'`,
-		schema, table, strings.Join(cols, ", "), file, r.s3Info.Region, delimiter)
-	opts := " IGNOREHEADER 0 ACCEPTINVCHARS TRUNCATECOLUMNS TRIMBLANKS BLANKSASNULL EMPTYASNULL DATEFORMAT 'auto' ACCEPTANYDATE COMPUPDATE ON"
-	creds := fmt.Sprintf(` CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'`, r.s3Info.AccessID, r.s3Info.SecretKey)
-	fullCopyCmd := fmt.Sprintf("%s%s%s", copyCmd, opts, creds)
-	_, err := r.logAndExec(fullCopyCmd)
-	return err
+		`COPY "%s"."%s" (%s) FROM '%s' WITH REGION '%s' %s CSV DELIMITER '%c'`,
+		schema, table, strings.Join(colStrings, ", "), file, r.s3Info.Region, gzipSQL, delimiter)
+	opts := "IGNOREHEADER 0 ACCEPTINVCHARS TRUNCATECOLUMNS TRIMBLANKS BLANKSASNULL EMPTYASNULL DATEFORMAT 'auto' ACCEPTANYDATE COMPUPDATE ON"
+	fullCopyCmd := fmt.Sprintf("%s %s %s", copyCmd, opts, credSQL)
```

Comment:
Again, I'd love to see this match the style of GetJSONCopySQL, or vice versa, rather than having one split out the cmd, opts, and creds, and pass them differently.

Also, if creds are just pasted into the end always, if we keep it separate, we don't even have to pass them into this routine, and then can more easily supply them as a separate argument to logAndExec.